ext/src/ruby_api/component/convert.rs

```diff
@@ -164,9 +181,64 @@ pub(crate) fn rb_to_component_val(
                 )?))))
             }
         }
-        Type::Result(_result_type) => not_implemented!("Result not implemented"),
+        Type::Result(result_type) => {
+            // Expect value to conform to `Wasmtime::Component::Value`'s interface
+            let ruby = Ruby::get().unwrap();
+            let is_ok = value.funcall::<_, (), bool>(IS_OK.into_id_with(&ruby), ())?;
+
+            if is_ok {
+                let ok_value = value.funcall::<_, (), Value>(OK.into_id_with(&ruby), ())?;
+                match result_type.ok() {
+                    Some(ty) => rb_to_component_val(ok_value, _store, &ty)
+                        .map(|val| Val::Result(Result::Ok(Some(Box::new(val))))),
+                    None => {
+                        if ok_value.is_nil() {
+                            Ok(Val::Result(Ok(None)))
+                        } else {
+                            err!(
+                                "expected nil for result<_, E> ok branch, got {}",
+                                ok_value.inspect()
+                            )
+                        }
+                    }
+                }
+            } else {
+                let err_value = value.funcall::<_, (), Value>(ERROR.into_id_with(&ruby), ())?;
+                match result_type.err() {
+                    Some(ty) => rb_to_component_val(err_value, _store, &ty)
+                        .map(|val| Val::Result(Result::Err(Some(Box::new(val))))),
+                    None => {
+                        if err_value.is_nil() {
+                            Ok(Val::Result(Err(None)))
+                        } else {
+                            err!(
+                                "expected nil for result<O, _> error branch, got {}",
+                                err_value.inspect()
+                            )
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
This is a mouthful. I tried simplifying but did not come up with anything much better. Ideas are welcome.